mitodl/mit-learn
# Fixes to address scim-form-keycloak weirdness

### What are the relevant tickets?




Part of https://github.com/mitodl/hq/issues/5561

### Description (What does it do?)

This makes some tweaks to the SCIM adaptor to handle some weird data we're getting from the scim-for-keycloak plugin I didn't previously notice:

- The plugin is embedding JSON strings inside the JSON requests, which we have to parse because otherwise the data is being ignored.
- The was one edge case where we weren't handling `emailOptIn`.
- I had to switch over using `displayName` to using `fullName`. `displayName` worked on full data syncs, but if you only updated your full name on the account page it wasn't triggering a SCIM PATCH request over to Learn. There must be some special handling of it in the plugin, despite the appearance in the admin UI that you could use `displayName` for a custom user property.

### How can this be tested?

- You'll need to update your SCIM Remote Provider User schema configuration by removing the custom attribute setting for `displayName` and adding a new attribute named `fullName` that uses the `fullName` custom attribute. All other settings can be left as the default.

You should be able to go through the following flows:

- Register as a new user:
  - Go to `/realms/olapps/account/` on Keycloak, this redirects to the login pages and avoids us exercising the social auth code on Learn that can obfuscate whether SCIM worked correctly or not.
  - Create a new user, confirm your email. You should now be on the profile page.
  - Go to `/admin` on the Learn API and verify the user has been created and has the correct name and the `email_optin` setting is set to true/yes.
- Update an existing user:
  - Go to `/realms/olapps/account/` on Keycloak and login as an existing user
  - Update combinations of the fields and confirm you see these updates propagate to Learn (e.g. test w/ updating all fields, test updating them individually, etc).
